ledger/participant-state/kvutils/tools/src/main/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityChecker.scala

```diff
@@ -94,40 +94,55 @@ class IntegrityChecker[LogResult](commitStrategySupport: CommitStrategySupport[L
     }
   }
 
-  private def compareWriteSets(expectedWriteSet: WriteSet, actualWriteSet: WriteSet): Unit = {
+  private def compareWriteSets(expectedWriteSet: WriteSet, actualWriteSet: WriteSet): Unit =
     if (expectedWriteSet == actualWriteSet) {
       println("OK".green)
     } else {
-      println("FAIL".red)
-      val message =
+      val messageMaybe =
         if (expectedWriteSet.size == actualWriteSet.size) {
-          expectedWriteSet
+          val differencesExplained = expectedWriteSet
             .zip(actualWriteSet)
-            .flatMap {
+            .map {
               case ((expectedKey, expectedValue), (actualKey, actualValue)) =>
                 if (expectedKey == actualKey && expectedValue != actualValue) {
-                  Seq(
-                    s"expected value:    ${bytesAsHexString(expectedValue)}",
-                    s" vs. actual value: ${bytesAsHexString(actualValue)}",
-                    explainDifference(expectedKey, expectedValue, actualValue),
-                  )
+                  explainDifference(expectedKey, expectedValue, actualValue).map {
+                    explainedDifference =>
+                      Seq(
+                        s"expected value:    ${bytesAsHexString(expectedValue)}",
+                        s" vs. actual value: ${bytesAsHexString(actualValue)}",
+                        explainedDifference,
+                      )
+                  }
                 } else if (expectedKey != actualKey) {
-                  Seq(
-                    s"expected key:    ${bytesAsHexString(expectedKey)}",
-                    s" vs. actual key: ${bytesAsHexString(actualKey)}",
-                  )
-                } else
-                  Seq.empty
+                  Some(
+                    Seq(
+                      s"expected key:    ${bytesAsHexString(expectedKey)}",
+                      s" vs. actual key: ${bytesAsHexString(actualKey)}",
+                    ))
+                } else {
+                  None
+                }
             }
+            .map(_.toList)
+            .filterNot(_.isEmpty)
+            .map(_.mkString(System.lineSeparator()))
             .mkString(System.lineSeparator())
```

Comment:
I don't think this does what you think it does.

```scala
scala> Seq(Some(Seq(1, 2)), None, Some(Seq(3)), Some(Seq(4, 5)), None, Some(Seq(6, 7, 8)), None).map(_.toList).filterNot(_.isEmpty).map(_.mkString(System.lineSeparator())).mkString(System.lineSeparator())
res0: String =
List(1, 2)
List(3)
List(4, 5)
List(6, 7, 8)
```

You have an extra level of nesting which is masked by the first `mkString`, because `toString` is unsafe. You need to flatten twice.

```scala
scala> Seq(Some(Seq(1, 2)), None, Some(Seq(3)), Some(Seq(4, 5)), None, Some(Seq(6, 7, 8)), None).map(_.toList).filterNot(_.isEmpty).flatten.flatten.mkString(System.lineSeparator())
res9: String =
1
2
3
4
5
6
7
8
```

I may be missing something, but I think you need to test this more thoroughly. Perhaps this part of the integrity checker needs some unit tests.

```suggestion
            .map(_.toList)
            .filterNot(_.isEmpty)
            .flatten
            .flatten
            .mkString(System.lineSeparator())
```

Reply:
Good point -- I'll add some tests for this bit.

Reply:
Added tests for it and fixed according to your suggestion. For multiple differences it was generating the wrong explanation string.